src/appleseed/renderer/modeling/aov/cryptomatte.cpp

```diff
@@ -0,0 +1,752 @@
+    sprintf(type_name_hex, "%08x", type_name_hash);
+
+    ImageAttributes image_attributes = ImageAttributes::create_default_attributes();
+    image_attributes.insert("color_space", "linear");
+
+    string layer_prefix("cryptomatte/");
+    layer_prefix.append(string(type_name_hex).erase(7));
+    image_attributes.insert(layer_prefix + "/conversion", "uint32_to_float32");
+    image_attributes.insert(layer_prefix + "/hash", "MurmurHash3_32");
+    image_attributes.insert(layer_prefix + "/name", layer_name);
+
+    NameMap name_map;
+    for (auto& pair : impl->m_tile_name_map)
+    {
+        for (auto& hash : pair.second)
+        {
+            if (hash.first == 0)
+                continue;
+            name_map[hash.first] = hash.second;
+        }
+    }
+
+    string manifest_str;
+    for (auto& hash : name_map)
+    {
+        char hash_hex[9];
+        sprintf(hash_hex, "%08x", hash.first);
+        manifest_str.append(format("\"{0}\":\"{1}\",", hash.second, string(hash_hex)));
+    }
+    if (manifest_str.size() > 0)
+        manifest_str.pop_back();
+
+    const string manifest_string = format("{{0}}", manifest_str);
+
+    image_attributes.insert(layer_prefix + "/manifest", manifest_string.c_str());
+
+    vector<string> channel_names;
+    channel_names.push_back("R");
+    channel_names.push_back("G");
+    channel_names.push_back("B");
+
+    int num_exr_layers = truncate<int>(fast_ceil(impl->m_num_layers / 2.0));
+    for(size_t i = 0; i < num_exr_layers; ++i)
+    {
+        string layer_number = get_numbered_string("##", i);
+        channel_names.push_back(format("{0}{1}.R", layer_name, layer_number));
+        channel_names.push_back(format("{0}{1}.G", layer_name, layer_number));
+        channel_names.push_back(format("{0}{1}.B", layer_name, layer_number));
+        channel_names.push_back(format("{0}{1}.A", layer_name, layer_number));
+    }
+
+    const Image& image = *impl->m_image;
+
+    try
+    {
+        const int num_channels = (impl->m_num_layers * 2) + 3;
+        MultiChannelExrFileWriter writer(exr_file_path.c_str(), &image);
+        writer.set_image_channels(num_channels, channel_names);
+        writer.set_image_attributes(image_attributes);
+        writer.write();
+    }
+    catch (const exception& e)
+    {
+        RENDERER_LOG_ERROR(
+            "failed to write cryptomatte file %s: %s.",
+            exr_file_path.c_str(),
+            e.what());
+
+        return false;
+    }
+
+    return true;
+}
+
+
+//
+// CryptomatteAOVFactory class implementation.
+//
+
+void CryptomatteAOVFactory::release()
+{
+    delete this;
+}
+
+const char* CryptomatteAOVFactory::get_model() const
+{
+    return CryptomatteAOVModel;
+}
+
+Dictionary CryptomatteAOVFactory::get_model_metadata() const
+{
+    return
+        Dictionary()
+        .insert("name", get_model())
+        .insert("label", "Cryptomatte");
+}
+
+DictionaryArray CryptomatteAOVFactory::get_input_metadata() const
+{
+    DictionaryArray metadata;
```

Comment:
We can add here the option to choose between object names or material names.
We don't have a way to edit them in studio but we'll have it.

Another option would be to have 2 Cryptomatte AOVs. One for object IDs and one for material IDs.
This option might be easier to support in plugins. I'll check that.

Reply:
I thought about two AOVs too, it will require another line `register_factory(auto_release_ptr<FactoryType>(new CryptomatteAOVFactory()));`
But mode parameter can passed to Factory class maybe?